servers/zms/src/main/java/com/yahoo/athenz/zms/store/impl/jdbc/JDBCConnection.java

```diff
@@ -6015,11 +6026,15 @@ public boolean insertRoleTags(String roleName, String domainName, Map<String, St
             throw notFoundError(caller, ZMSConsts.OBJECT_ROLE, ResourceUtils.roleResourceName(domainName, roleName));
         }
         int curTagCount = getRoleTagsCount(roleId);
-
-        int remainingTagsToInsert = roleTagsLimit - curTagCount;
+        int newTagCount = calculateTagCount(roleTags);
+        if (curTagCount + newTagCount > roleTagsLimit) {
+            throw requestError(caller, "role tag quota exceeded - limit: "
+                + roleTagsLimit + ", current tags count: " + curTagCount + ", new tags count: " + newTagCount);
+        }
+        
         boolean res = true;
         for (Map.Entry<String, StringList> e : roleTags.entrySet()) {
-            for (int i = 0; i < e.getValue().getList().size() && remainingTagsToInsert-- > 0; i++) {
+            for (int i = 0; i < e.getValue().getList().size(); i++) {
```

Comment:
since we no longer have the extra condition in the for loop, we can replace this with a simplified for loop:
for (String tagValue : e.getValue().getList())